nixos/apple-touchbar: init module

andre4ik3 opened the pull request:
This PR adds `hardware.apple.touchBar`, a module for setting up the Touch Bar on some Apple laptops. It uses [tiny-dfr](https://github.com/WhatAmISupposedToPutHere/tiny-dfr), a small daemon originating out of the Asahi Linux project (but works also on T2 Macs, already in nixpkgs).

With this change, enabling support for the Touch Bar becomes one line of configuration:

```
hardware.apple.touchBar.enable = true;
```

followed by a rebuild and reboot. Tiny DFR can also be configured through this module, like so:

```
hardware.apple.touchBar.settings = {
  MediaLayerDefault = true;
  ShowButtonOutlines = false;
  EnablePixelShift = true;
};
```

##### Reviewed points

- [x] module path fits the guidelines
  - see comments
- [ ] module tests, if any, succeed on ARCHITECTURE
- [ ] module tests, if any, are added to package `passthru.tests`
- [x] options have appropriate types
- [x] options have default
- [x] options have example
- [x] options have descriptions
- [x] No unneeded package is added to `environment.systemPackages`
- [ ] `meta.maintainers` is set
- [ ] module documentation is declared in `meta.doc`

##### Possible improvements

- ~~Service can be restarted after config changes~~ done

##### Comments

- I put it as `hardware.apple.touchBar` instead of e.g. `services.tiny-dfr` because of the following:
  - Since it's hardware dependent and related to setting up hardware, I think it should go in `hardware`.
  - Touch Bar only exists on Apple devices, hence `hardware.apple`. This is more descriptive than just simply `hardware.touchBar` and as a bonus, creates a space for any future Apple hardware modules to go into.
  - From my understanding, `tiny-dfr` is the main implementation of support for Touch Bars. The only other implementation I'm aware of is `touchbard` and it's deprecated, with `tiny-dfr` being the recommended replacement.
  - I put `touchBar` instead of `touchbar` as in official branding, documentation etc. it's "Touch Bar", two words, so translating to camelCase it becomes `touchBar`.
- I'm not sure if I should put myself as maintainer, as it's really quite a simple module and I didn't write any of the code for it, I just wrote this very simple integration. Previously I did it a more complicated way by manually doing the udev rules and systemd units, but now I've simplified it to just use the rules and units provided by the package.
- All of the above is just my opinion, I'm more than happy to change anything, I'm new to contributing to NixOS and am just trying to upstream some stuff I've had privately for a while.

---

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

andre4ik3 committed:
Initial attempt at Touch Bar support using tiny-dfr

andre4ik3 committed:
Add apple-touchbar to modules list

andre4ik3 committed:
Use the udev rules in tiny-dfr package instead of copying them

andre4ik3 committed:
Use systemd units from tiny-dfr package as well

andre4ik3 committed:
Add configuration example for settings

andre4ik3 committed:
Add changelog

andre4ik3 committed:
Fix formatting

SuperSandro2000 commented on nixos/modules/hardware/apple-touchbar.nix:
```diff
@@ -0,0 +1,48 @@
+  ...
+}:
+
+with lib;
```
```suggestion
```
we don't want to do that anymore

andre4ik3 replied:
Fixed

andre4ik3 committed:
Stop using with lib

andre4ik3 committed:
Add restart trigger

Normally tiny-dfr watches for config file changes, but `system.etc.overlay.enable` seems to break it

andre4ik3 committed:
Add package override option, further simplify
